:book: Add the quickstart details for KubeVirt (#7444)
kubernetes-sigs/cluster-api: docs/book/src/user/quick-start.md

   Then follow the instruction for your kind version using  `kind create cluster --config kind-cluster-with-extramounts.yaml`
   to create the management cluster using the above file.

   {{#/tab }}
   {{#tab KubeVirt}}

   #### Create the Kind Cluster

>    To allow using LoadBalancer, we can't use the kind's default CNI (kindnet), but we'll need to install another CNI, like Calico. In order to do that, we'll need first to initiate the kind cluster with two modifications:

Reviewer: I think it could be useful to add a note explaining why it is required to have a load balancer in the management cluster (which is something we do not have for other providers)?

   1. Disable the default CNI
   2. Add the docker credentials to the cluster, to avoid the docker hub pull rate limit.
   
   Create a configuration file for kind. Please notice the docker config file path, and adjust it to your local setting:
   ```bash
   cat <<EOF > kind-config.yaml
   kind: Cluster
   apiVersion: kind.x-k8s.io/v1alpha4
   networking:
   # the default CNI will not be installed
     disableDefaultCNI: true
   nodes:
   - role: control-plane
     extraMounts:
      - containerPath: /var/lib/kubelet/config.json
        hostPath: <YOUR DOCKER CONFIG FILE PATH>
   EOF
   ```
   Now, create the kind cluster with the configuration file:
   ```bash
   kind create cluster --config=kind-config.yaml
   ```
   Test to ensure the local kind cluster is ready:
   ```bash
   kubectl cluster-info
   ```

   #### Install the Calico CNI
   Download the Calico manifest and use it to create the required resources, as described [here](https://projectcalico.docs.tigera.io/getting-started/kubernetes/self-managed-onprem/onpremises#install-calico); e.g.:
   ```bash

>    curl https://raw.githubusercontent.com/projectcalico/calico/v3.24.3/manifests/calico.yaml -O

Reviewer 2: Do we want to pin Calico to a specific version? It will require frequent updates to the book in order to not get outdated which might make this doc harder to maintain

Reviewer 3: In the "others..." tab it has been proven to be sustainable as we usually bump calico as well when we bump Kubernetes.

I don't know how sustainable it is for kubevirt (as we wouldn't bump that when we bump for others as we don't have the time to test it).

We pin in the "others..." tab to ensure we have something which has been tested manually at least once. We couldn't guarantee this with some sort of "latest".

Author: Thanks for your comment @CecileRobertMichon. I was wandering about it when I wrote it. But this is not only a matter of versions, because new calico version my also bring relevant API changes. So I prefer to pin it for now, and monitor it. Any change will require some manual testing (i.e. go over the whole guide again and again to make sure everything is still working).

Reviewer 2: That's fair 👍 SGTM

   Then follow the instruction for your kind version using  `kind create cluster --config kind-cluster-with-extramounts.yaml`
   to create the management cluster using the above file.

   {{#/tab }}
   {{#tab KubeVirt}}

   #### Create the Kind Cluster
   [KubeVirt][KubeVirt] is a cloud native virtualization solution. The virtual machines we're going to create and use for
   the workload cluster's nodes, are actually running within pods in the management cluster. In order to communicate
   the workload cluster's API server, we'll need to expose it. We are using Kind which is a limited environment. The
   easiest way to expose the workload cluster's API server (a pod within a node running in a VM that is itself running
   within a pod in the management cluster, that is running inside a docker container), is to use a LoadBalancer service.

   To allow using a LoadBalancer service, we can't use the kind's default CNI (kindnet), but we'll need to install
   another CNI, like Calico. In order to do that, we'll need first to initiate the kind cluster with two modifications:
   1. Disable the default CNI

>    2. Add the docker credentials to the cluster, to avoid the docker hub pull rate limit.

Reviewer 4: How frequently does this happen? Are the kubevirt images in docker hub?
Asking because if it's not always necessary, we can omit it and maybe add it as troubleshooting step/advice. It adds one more thing the user needs to do and since this is a quick start, the less steps, the better.

Author: @g-gaston, this is pretty frequent. I would leave it here because (a) we need the configuration file anyway to disable kindnet CNI, and (b) because I think it's more than a regular troubleshoot, because starting the kind cluster is a basic pre-requisite. To re-do it, you must start over the whole process. So I think we need the right setting in advance.

   Create a configuration file for kind. Please notice the docker config file path, and adjust it to your local setting:
   ```bash
   cat <<EOF > kind-config.yaml
   kind: Cluster
   apiVersion: kind.x-k8s.io/v1alpha4
   networking:
   # the default CNI will not be installed
     disableDefaultCNI: true
   nodes:
   - role: control-plane
     extraMounts:
      - containerPath: /var/lib/kubelet/config.json

>         hostPath: <YOUR DOCKER CONFIG FILE PATH>

Reviewer 4: If you want to keep this because the rate limit happens frequently, I would suggest to add a bit more info on how to create/find this file, or maybe just link to the official docs if there is any.

One could argue this is pretty basic and any CAPI user should already know how to do it. However, this is a quick start guide, so the lower the barrier of entry, the better :)

   EOF
   ```
   Now, create the kind cluster with the configuration file:
   ```bash
   kind create cluster --config=kind-config.yaml
   ```
   Test to ensure the local kind cluster is ready:
   ```bash
   kubectl cluster-info
   ```

   #### Install the Calico CNI
   Now we'll need to install a CNI. In this example, we're using calico, but other CNIs should work as well. Please see
   [calico installation guide](https://projectcalico.docs.tigera.io/getting-started/kubernetes/self-managed-onprem/onpremises#install-calico)
   for more details (use the "Manifest" tab). Below is an example of how to install calico version v3.24.4.

   Download the Calico manifest and use it to create the required resources; e.g.:
   ```bash

>    curl https://raw.githubusercontent.com/projectcalico/calico/v3.24.4/manifests/calico.yaml -O
>    kubectl create -f calico.yaml

Reviewer 4: nit: you could make this easier
```suggestion
   kubectl create -f https://raw.githubusercontent.com/projectcalico/calico/v3.24.4/manifests/calico.yaml
```

{{#/tab }}
{{#tab Kubevirt}}

Please visit the [Kubevirt project][Kubevirt provider] for more information.

KubeVirt is a cloud native virtualization solution. The meaning for that for our guest cluster, is that the virtual

> machines will be created within the kind cluster, as a kubernetes resources. To make it work, we'll need to add some

Reviewer 3: ```suggestion
machines will be created within the kind cluster, as a Kubernetes resources. To make it work, we'll need to add some
```

configurations to our kind cluster.

> ##### Install the calico CNI

Reviewer 3: ```suggestion
##### Install the Calico CNI
```

Reviewer 3: (same in other places)

> The default kind's CNI (kindnet) will not work for the guest cluster. We'll use calico CNI. Download the calico manifest and use it to create the required resources, as described [here](https://projectcalico.docs.tigera.io/getting-started/kubernetes/self-managed-onprem/onpremises#install-calico); e.g.:

Reviewer 3: ```suggestion
kind's default CNI (kindnet) will not work for the guest cluster. We'll use calico CNI. Download the calico manifest and use it to create the required resources, as described [here](https://projectcalico.docs.tigera.io/getting-started/kubernetes/self-managed-onprem/onpremises#install-calico); e.g.:
```

```bash
curl https://raw.githubusercontent.com/projectcalico/calico/v3.24.3/manifests/calico.yaml -O

> kubectl create -f calico.yaml

Reviewer 3: Does this mean that after that kindnet & Calico are installed? Isn't that a problem?

Author: > Does this mean that after that kindnet & Calico are installed? Isn't that a problem?

This is a good question, @sbueringer. So in my environment it's working (tested over and over again). It is possible to disable kindnet (https://kind.sigs.k8s.io/docs/user/configuration/#disable-default-cni) but I thought it will make the document a bit too complicated, because I'll need a custom kind cluster creation for KubeVirt, and since it does work without this modification, I thought it will be simpler to leave it as is. WDYT?

```

> ##### Install metallb for load balancing

Reviewer 3: ```suggestion
##### Install MetalLB for load balancing
```

```

{{#/tab }}
{{#tab KubeVirt}}

Please visit the [KubeVirt project][KubeVirt provider] for more information.

KubeVirt is a cloud native virtualization solution that allows running and managing virtual machines within Kubernetes 
clusters. This means we can use kind cluster to both host the cluster api components within pods and the guest cluster 
within KubeVirt virtual machines. In order to make this work, we need to apply some configurations to our kind cluster.

#### Install MetalLB for load balancing

> We'll need to enable load-balance services in our kind cluster. For that, we'll install MetalLB, as describe [here](https://metallb.universe.tf/installation/#installation-by-manifest); e.g.

Reviewer: nit
```suggestion
We'll need to enable load-balance services in our kind cluster. For this tutorial, we'll install MetalLB, as described [here](https://metallb.universe.tf/installation/#installation-by-manifest); e.g.
```
Assuming that any LB could work

```

{{#/tab }}
{{#tab KubeVirt}}

Please visit the [KubeVirt project][KubeVirt provider] for more information.

As described above, we want to use a LoadBalancer service in order to expose the workload cluster's API server. In the
example below, we will use [MetalLB](https://metallb.universe.tf/) solution to implement load balancing to our kind
cluster. Other solution should work as well.

#### Install MetalLB for load balancing
Install MetalLB, as described [here](https://metallb.universe.tf/installation/#installation-by-manifest); for example:
```bash
METALLB_VER=$(curl "https://api.github.com/repos/metallb/metallb/releases/latest" | jq -r ".tag_name")
kubectl apply -f "https://raw.githubusercontent.com/metallb/metallb/${METALLB_VER}/config/manifests/metallb-native.yaml"
kubectl wait pods -n metallb-system -l app=metallb,component=controller --for=condition=Ready --timeout=10m
kubectl wait pods -n metallb-system -l app=metallb,component=speaker --for=condition=Ready --timeout=2m
```

> Now, we'll Create the `IPAddressPool` and the `L2Advertisement` custom resources. The script below creates the CRs with

Reviewer 3: ```suggestion
Now, we'll create the `IPAddressPool` and the `L2Advertisement` custom resources. The script below creates the CRs with
```
nit

Author: Thanks @sbueringer - done

Reviewer 3: Thx!

{{#/tab }}
{{#tab Kubevirt}}

```bash
export CAPK_GUEST_K8S_VERSION="${CAPK_GUEST_K8S_VERSION:-v1.23.10}"
export CRI_PATH="${CRI_PATH:-/var/run/containerd/containerd.sock}"
export NODE_VM_IMAGE_TEMPLATE="${NODE_VM_IMAGE_TEMPLATE:-quay.io/capk/ubuntu-2004-container-disk:${CAPK_GUEST_K8S_VERSION}}"

> export IMAGE_REPO="${IMAGE_REPO:-k8s.gcr.io}"

Reviewer 3: fyi. The new default with v1.25 is registry.k8s.io (there's a redirect right now, so both serve the same images)

Please consider using the new registry

Author: The latest image is for 1.23

@davidvossel - are we planing to introduce an image for 1.25?

Author: For some reason this is just not working and I can't debug it. I will leave the original value of `k8s.gcr.io`.

Reviewer 3: Okay, please follow-up on this. Kubernetes is literally running out of money with the old registry. So we're trying everything we can going forward to use the new registry as it's cheaper.

Please visit the [IBM Cloud provider] for more information.

{{#/tab }}
{{#tab KubeVirt}}

```bash
export CAPK_GUEST_K8S_VERSION="v1.23.10"
export CRI_PATH="/var/run/containerd/containerd.sock"
export NODE_VM_IMAGE_TEMPLATE="quay.io/capk/ubuntu-2004-container-disk:${CAPK_GUEST_K8S_VERSION}"

> export IMAGE_REPO="k8s.gcr.io"

Reviewer: can we switch this to the new registry, images should be already available in both (to be verified)

Author: I tried that several times and it's just not working. After looking at the code, I think it should only work from K8s v1.25+.

I also noticed that other providers just dropped this setting, so I opened a PR to drop it also in Kubevirt provider (kubernetes-sigs/cluster-api-provider-kubevirt#207) and I hope it will be included in the next version.

Reviewer 3: We are using the new registry for Kubernetes >= 1.22

Reviewer 3: +1 to your PR. I would definitely recommend not exposing IMAGE_REPO as an env var

    --target-namespace ${CLUSTER_NAMESPACE} | kubectl apply -f -
```

{{#/tab }}
{{#tab kubevirt}}

> First, download the cluster manifest, because we'll need to modify it a bit:

Reviewer 3: ```suggestion
First, download the cluster template, because we'll need to modify it a bit:
```

```bash
CAPK_VER=$(curl "https://api.github.com/repos/kubernetes-sigs/cluster-api-provider-kubevirt/releases/latest" | jq -r ."tag_name")
curl -L https://github.com/kubernetes-sigs/cluster-api-provider-kubevirt/releases/download/${CAPK_VER}/cluster-template.yaml -O
```
Now, modify the service to be with the`LoadBalancer` type:
```bash

> sed -i 's/type: ClusterIP/type: LoadBalancer/' cluster-template.yaml

Reviewer 3: Would it be possible to do this change to cluster-template in kubevirt itself?

Would be easier for users vs. telling them to modify the released cluster-template

Author: @davidvossel - what do you think?

Author: OK - working on it, here: https://github.com/kubernetes-sigs/cluster-api-provider-kubevirt/pull/202

    --target-namespace ${CLUSTER_NAMESPACE} | kubectl apply -f -
```

{{#/tab }}
{{#tab KubeVirt}}

Create the workload cluster using the load balancer template  
```bash
clusterctl generate cluster capi-quickstart \
  --infrastructure="kubevirt" \

>   --flavor lb \

Reviewer: is there a reason why we are not using the default template?

    --target-namespace ${CLUSTER_NAMESPACE} | kubectl apply -f -
```

{{#/tab }}
{{#tab KubeVirt}}

Create the workload cluster using the load balancer template  
```bash

> CAPK_VER=$(curl "https://api.github.com/repos/kubernetes-sigs/cluster-api-provider-kubevirt/releases/latest" | jq -r ."tag_name")
> 
> curl -L -O "https://github.com/kubernetes-sigs/cluster-api-provider-kubevirt/releases/download/${CAPK_VER}/cluster-template-lb.yaml"
> 
> clusterctl generate cluster capi-quickstart \
>   --kubernetes-version ${CAPK_GUEST_K8S_VERSION} \
>   --control-plane-machine-count=1 \
>   --worker-machine-count=1 \
>   --from cluster-template-lb.yaml \
>   > capi-quickstart.yaml

Reviewer 3: ```suggestion
clusterctl generate cluster capi-quickstart --flavor lb \
  --kubernetes-version ${CAPK_GUEST_K8S_VERSION} \
  --control-plane-machine-count=1 \
  --worker-machine-count=1 \
  --from cluster-template-lb.yaml \
  > capi-quickstart.yaml
```

Should work now

Author: :+1: :+1: :+1: 

These are good news. I opened a related issue (#7455) + PR to fix it (#7453), in order to use the `--form` flag to get a template from release assets. I still think it should be merged, but in this case, the `--flavor` flag will work.

Reviewer 3: Sounds good to me!

Author: @sbueringer - done.

Reviewer 3: Thx!

</aside>

> {{#/tab }}

Reviewer 3: Why was this dropped?

Author: ooops

Calico not required for vcluster.

{{#/tab }}

> {{#tab kubevirt}}

Reviewer 3: ```suggestion
{{#tab Kubevirt}}
```
(same in l.1204, 1041 and 1077)

Author: Changed all to `KubeVirt`

Before deploying the calico CNI, make sure the VMs are running:
```bash
kubectl get vm
```

If our new VMs are running, we should see a response similar to this:

```text
NAME                                  AGE    STATUS    READY
capi-quickstart-control-plane-7s945   167m   Running   True
capi-quickstart-md-0-zht5j            164m   Running   True
```

We can also read the virtual machine instances:
```bash
kubectl get vmi
```
The output will be similar to:
```text
NAME                                  AGE    PHASE     IP             NODENAME             READY
capi-quickstart-control-plane-7s945   167m   Running   10.244.82.16   kind-control-plane   True
capi-quickstart-md-0-zht5j            164m   Running   10.244.82.17   kind-control-plane   True
```

> We'll need to prevent conflicts with the kind cluster network, buy modifying the default calico settings:

Reviewer 3: ```suggestion
We'll need to prevent conflicts with the kind cluster network, by modifying the default calico settings:
```

> * change the CIDR to a non-conflict range

Reviewer 3: ```suggestion
* change the CIDR to a non-conflicting range
```

> * change the CIDR to a non-conflict range
> * Change the value of the `CLUSTER_TYPE` environment variable to `k8s`
> * Change the value of the `CALICO_IPV4POOL_IPIP` environment variable to `Never`
> * Change the value of the `CALICO_IPV4POOL_VXLAN` environment variable to `Always`
> * add the `FELIX_VXLANPORT` environment variable with the value of a non-conflict port, e.g. `"6789"`.

Reviewer 3: ```suggestion
* Change the CIDR to a non-conflict range
* Change the value of the `CLUSTER_TYPE` environment variable to `k8s`
* Change the value of the `CALICO_IPV4POOL_IPIP` environment variable to `Never`
* Change the value of the `CALICO_IPV4POOL_VXLAN` environment variable to `Always`
* Add the `FELIX_VXLANPORT` environment variable with the value of a non-conflicting port, e.g. `"6789"`.
```
nit

The following script downloads the calico manifest and modifies the required field. The CIDR and the port values are examples.
```bash
curl https://raw.githubusercontent.com/projectcalico/calico/v3.24.3/manifests/calico.yaml -O

sed -i -E 's|^( +)# (- name: CALICO_IPV4POOL_CIDR)$|\1\2|g;'\
's|^( +)# (  value: )"192.168.0.0/16"|\1\2"10.243.0.0/16"|g;'\
'/- name: CLUSTER_TYPE/{ n; s/( +value: ").+/\1k8s"/g };'\
'/- name: CALICO_IPV4POOL_IPIP/{ n; s/value: "Always"/value: "Never"/ };'\
'/- name: CALICO_IPV4POOL_VXLAN/{ n; s/value: "Never"/value: "Always"/};'\
'/# Set Felix endpoint to host default action to ACCEPT./a\            - name: FELIX_VXLANPORT\n              value: "6789"' \
calico.yaml
```

> Now, deploy the calico CNI on the guest cluster:

Reviewer 3: ```suggestion
Now, deploy the calico CNI on the workload cluster:
```
Should we call it workload cluster? (given that's how CAPI calls it in general)

(also in other places)

Calico not required for vcluster.

{{#/tab }}
{{#tab KubeVirt}}

Before deploying the Calico CNI, make sure the VMs are running:
```bash
kubectl get vm
```

If our new VMs are running, we should see a response similar to this:

```text
NAME                                  AGE    STATUS    READY
capi-quickstart-control-plane-7s945   167m   Running   True
capi-quickstart-md-0-zht5j            164m   Running   True
```

We can also read the virtual machine instances:
```bash
kubectl get vmi
```
The output will be similar to:
```text
NAME                                  AGE    PHASE     IP             NODENAME             READY
capi-quickstart-control-plane-7s945   167m   Running   10.244.82.16   kind-control-plane   True
capi-quickstart-md-0-zht5j            164m   Running   10.244.82.17   kind-control-plane   True
```

> We'll need to prevent conflicts with the kind cluster network, by modifying the default Calico settings:

Reviewer: I think it will be beneficial for users to explain why we need this change

* Change the CIDR to a non-conflicting range
* Change the value of the `CLUSTER_TYPE` environment variable to `k8s`
* Change the value of the `CALICO_IPV4POOL_IPIP` environment variable to `Never`
* Change the value of the `CALICO_IPV4POOL_VXLAN` environment variable to `Always`
* Add the `FELIX_VXLANPORT` environment variable with the value of a non-conflicting port, e.g. `"6789"`.

The following script downloads the Calico manifest and modifies the required field. The CIDR and the port values are examples.
```bash

> curl https://raw.githubusercontent.com/projectcalico/calico/v3.24.3/manifests/calico.yaml -o calico-workload.yaml

Reviewer 2: same question here about calico version
